engine/src/main/scala/cromwell/engine/workflow/WorkflowDockerLookupActor.scala

```diff
@@ -0,0 +1,322 @@
+
+  context.become(dockerReceive orElse receive)
+
+  {
+    val restart = startMode == RestartExistingWorkflow
+    val initialState = if (restart) AwaitingFirstRequestOnRestart else Running
+
+    startWith(initialState, WorkflowDockerLookupActorData.empty)
+  }
+
+  // `AwaitingFirstRequestOnRestart` is only used in restart scenarios.  This state waits until there's at least one hash
+  // request before trying to load the docker hash mappings.  This is so we'll have at least one `JobPreparationActor`
+  // reference available to message with a terminal failure in case the reading or parsing of these mappings fails.
+  when(AwaitingFirstRequestOnRestart) {
+    case Event(request: DockerHashRequest, _) =>
+      loadDockerHashStoreEntries()
+      self forward request
+      goto(LoadingCache)
+  }
+
+  // Waiting for a response from the database with the hash mapping for this workflow.
+  when(LoadingCache) {
+    case Event(request: DockerHashRequest, data) =>
+      stay using data.addHashRequest(request, sender())
+    case Event(DockerHashStoreLoadingSuccess(dockerHashEntries), data) =>
+      loadCacheAndHandleHashRequests(dockerHashEntries, data)
+  }
+
+  // This is the normal operational mode.
+  when(Running) {
+    // This tag has already been looked up and its hash is in the mappings cache.
+    case Event(request: DockerHashRequest, data) if data.mappings.contains(request.dockerImageID) =>
+      sender ! DockerHashSuccessResponse(data.mappings(request.dockerImageID), request)
+      stay()
+    // A request for the hash for this tag has already been made to the hashing actor.  Don't request the hash again,
+    // just add this sender to the list of replyTos for when the hash arrives.
+    case Event(request: DockerHashRequest, data) if data.hashRequests.contains(request) =>
+      stay using data.addHashRequest(request, sender())
+    // This tag has not (successfully) been looked up before, so look it up now.
+    case Event(request: DockerHashRequest, data) =>
+      requestDockerHash(request, data)
+    case Event(dockerResponse: DockerHashSuccessResponse, data) =>
+      persistDockerHash(dockerResponse, data)
+    case Event(dockerResponse: DockerHashFailureResponse, data) =>
+      handleLookupFailure(dockerResponse, data)
+    case Event(DockerHashStoreSuccess(response), data) =>
+      recordMappingAndRespond(response, data)
+    case Event(DockerHashStoreFailure(request, e), data) =>
+      handleStoreFailure(request, new Exception(s"Failure storing docker hash for ${request.dockerImageID.fullName}", e), data)
+  }
+
+  // In state Terminal we reject all requests with the cause set in the state data.
+  when(Terminal) {
+    case Event(request: DockerHashRequest, data) =>
+      sender() ! WorkflowDockerLookupFailure(data.failureCause.orNull, request)
+      stay()
+  }
+
+  private def fail(reason: Throwable): State = {
+    self ! TransitionToFailed(reason)
```

Comment:
Some of the failures can occur in callbacks.  I wanted to route all failures through one path, which as you point out does result in some otherwise unnecessary messages to self.